"Look Up" and "Copy Subject" context menu items are added after the menu is visible, which changes the current item under the cursor

rr-codes opened the pull request:
#### 4d608c217abcf334bf7ffd7b0848dece20489494
<pre>
&quot;Look Up&quot; and &quot;Copy Subject&quot; context menu items are added after the menu is visible, which changes the current item under the cursor
<a href="https://bugs.webkit.org/show_bug.cgi?id=267138">https://bugs.webkit.org/show_bug.cgi?id=267138</a>
<a href="https://rdar.apple.com/111049353">rdar://111049353</a>

Reviewed by Aditya Keerthi.

Show the items as disabled instead of hiding the &quot;Copy Subject&quot; and &quot;Look Up&quot; items until they are
ready to be used.

Note that this has the consequence that these items may never change from the disabled state to the
enabled state if it turns out an image isn&apos;t actually eligible for a given item.

* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::requestBackgroundRemoval):
* Source/WebKit/Shared/WebContextMenuItemData.h:
(WebKit::WebContextMenuItemData::setEnabled):
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(-[_WKElementAction _initWithTitle:actionHandler:type:assistant:disabled:]):
(+[_WKElementAction elementActionWithTitle:actionHandler:]):
(+[_WKElementAction _elementActionWithType:title:actionHandler:]):
(+[_WKElementAction _elementActionWithType:customTitle:assistant:disabled:]):
(+[_WKElementAction _elementActionWithType:info:assistant:]):
(+[_WKElementAction _elementActionWithType:info:assistant:disabled:]):
(+[_WKElementAction elementActionWithType:customTitle:]):
(-[_WKElementAction uiActionForElementInfo:]):
(-[_WKElementAction _initWithTitle:actionHandler:type:assistant:]): Deleted.
(+[_WKElementAction _elementActionWithType:customTitle:assistant:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKElementActionInternal.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForLinkSheet:]):
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _insertDynamicImageAnalysisContextMenuItemsIfPossible]):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getContextMenuFromItems):

Canonical link: <a href="https://commits.webkit.org/272869@main">https://commits.webkit.org/272869@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1580a7efe1985795fb0f00ab576bd8ef17fc9038

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33416 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/12187 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/35329 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/36042 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/30365 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14537 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/9329 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29496 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33889 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/10281 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29830 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8972 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9100 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37372 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30347 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/30166 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35220 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9223 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33088 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10975 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7732 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9808 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9947 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rr-codes commented on Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
```diff
@@ -323,7 +323,9 @@ void requestBackgroundRemoval(CGImageRef image, CompletionHandler<void(CGImageRe
         callOnMainRunLoop([protectedResult = RetainPtr { result }, completion = WTFMove(completion), startTime]() mutable {
             RELEASE_LOG(ImageAnalysis, "Remove background finished in %.0f ms (found subject? %d)", (MonotonicTime::now() - startTime).milliseconds(), !!protectedResult);
-            completion(protectedResult.get());
+            RunLoop::main().dispatchAfter(5_s, [protectedResult, completion = WTFMove(completion)]() mutable {
```
oops

aprotyas replied:
😅

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12280,15 +12280,29 @@ - (void)_insertDynamicImageAnalysisContextMenuItemsIfPossible
+            if (![child isKindOfClass:UIAction.class])
+                continue;
+
+            UIAction *action = (UIAction *)child;
```
Please use `dynamic_objc_cast`.

rr-codes replied:
clearly I have spent too much time recently in Obj-C land 😂

rr-codes commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -712,21 +712,26 @@ static void getStandardShareMenuItem(NSArray *items, void (^completionHandler)(N
 
 #if ENABLE(IMAGE_ANALYSIS)
-    filteredItems.removeAllMatching([&] (auto& item) {
+    for (auto& item in filteredItems) {
```
lol too much combining of Obj C and C++

aprotyas commented on Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
```diff
@@ -712,21 +712,26 @@ static void getStandardShareMenuItem(NSArray *items, void (^completionHandler)(N
+    for (auto& item in filteredItems) {
         switch (item.action()) {
-        case ContextMenuItemTagLookUpImage:
+        case ContextMenuItemTagLookUpImage: {
```
nit: Are the braces necessary?

aprotyas reviewed (commented):
(Looks good! Bar the Obj-C/Cpp loop mishmash)

pxlcoder commented on Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
```diff
@@ -561,7 +561,7 @@ - (void)_appendAnimationAction:(NSMutableArray *)actions elementInfo:(_WKActivat
 #if HAVE(SAFARI_SERVICES_FRAMEWORK)
     if ([getSSReadingListClass() supportsURL:targetURL])
-        [defaultActions addObject:[_WKElementAction _elementActionWithType:_WKElementActionTypeAddToReadingList info:elementInfo assistant:self]];
+        [defaultActions addObject:[_WKElementAction _elementActionWithType:_WKElementActionTypeAddToReadingList info:elementInfo assistant:self disabled:NO]];
```
Can't we just use `+(instancetype)_elementActionWithType:(_WKElementActionType)type info:(_WKActivatedElementInfo *)info assistant:(WKActionSheetAssistant *)assistant` here?

Same comment applies to a couple other places.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
```diff
@@ -102,7 +103,7 @@ - (void)dealloc
 {
     return adoptNS([[self alloc] _initWithTitle:title actionHandler:^(WKActionSheetAssistant *, _WKActivatedElementInfo *actionInfo) { handler(actionInfo); }
-        type:_WKElementActionTypeCustom assistant:nil]).autorelease();
+       type:_WKElementActionTypeCustom assistant:nil]).autorelease();
```
Nit: This looks wrong? It was a 4 character indent before.

pxlcoder commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -12280,15 +12280,29 @@ - (void)_insertDynamicImageAnalysisContextMenuItemsIfPossible
+                continue;
+            }
+
+            if ([action.identifier isEqual:elementActionTypeToUIActionIdentifier(_WKElementActionTypeRevealImage)]) {
```
I think you can reuse the local variable here: `revealImageIdentifier`.

rr-codes committed:
"Look Up" and "Copy Subject" context menu items are added after the menu is visible, which changes the current item under the cursor
https://bugs.webkit.org/show_bug.cgi?id=267138
rdar://111049353

Reviewed by Aditya Keerthi.

Show the items as disabled instead of hiding the "Copy Subject" and "Look Up" items until they are
ready to be used.

Note that this has the consequence that these items may never change from the disabled state to the
enabled state if it turns out an image isn't actually eligible for a given item.

* Source/WebKit/Platform/cocoa/ImageAnalysisUtilities.mm:
(WebKit::requestBackgroundRemoval):
* Source/WebKit/Shared/WebContextMenuItemData.h:
(WebKit::WebContextMenuItemData::setEnabled):
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(-[_WKElementAction _initWithTitle:actionHandler:type:assistant:disabled:]):
(+[_WKElementAction elementActionWithTitle:actionHandler:]):
(+[_WKElementAction _elementActionWithType:title:actionHandler:]):
(+[_WKElementAction _elementActionWithType:customTitle:assistant:disabled:]):
(+[_WKElementAction _elementActionWithType:info:assistant:]):
(+[_WKElementAction _elementActionWithType:info:assistant:disabled:]):
(+[_WKElementAction elementActionWithType:customTitle:]):
(-[_WKElementAction uiActionForElementInfo:]):
(-[_WKElementAction _initWithTitle:actionHandler:type:assistant:]): Deleted.
(+[_WKElementAction _elementActionWithType:customTitle:assistant:]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKElementActionInternal.h:
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForLinkSheet:]):
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _insertDynamicImageAnalysisContextMenuItemsIfPossible]):
* Source/WebKit/UIProcess/mac/WebContextMenuProxyMac.mm:
(WebKit::WebContextMenuProxyMac::getContextMenuFromItems):

Canonical link: https://commits.webkit.org/272869@main
